<a href="https://colab.research.google.com/github/cpalme28/MAT-421/blob/main/Module_C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Homework 4 - Coree Palmer

**19.1 - Root Finding Problem Statement**

The root or zero of a function, $f(x)$, is an $x_r$ such that $f(x_r) = 0$. For some functions, determining an analytic solution for the roots of functions can be difficult. In those cases, it is helpful to generate numerical approximations of the roots of f.

In [3]:
# compute the root of f(x) = cos(x) - x near -2

import numpy as np
from scipy import optimize

f = lambda x: np.cos(x) - x
r = optimize.fsolve(f, -2)
print("r = ", r)

# verification
result = f(r)
print("result = ", result)

r =  [0.73908513]
result =  [0.]


The result of the function is 0 because 0.73908513 is a root, showing that fsolve works properly.

The function f(x) = 1/x has no root, however. Using the fsolve function to compute it will showcase what will happen when there is no root.

In [8]:
f = lambda x: 1/x**3

r, infodict, ier, mesg = optimize.fsolve(f, -2, full_output=True)
print("r = ", r)

result = f(r)
print("result = ", result)

print(mesg)

r =  [-6.3981792e+34]
result =  [-3.81795496e-105]
The number of calls to function has reached maxfev = 400.


As the output shows, the value of r is not a root, despite f(r) being a very small number. Turning on full_output allows us to see more information, letting us understand more into the failure.

**19.2 - Tolerance**

Error is a deviation from an expected or computed value. Tolerance is the level of error that is acceptable for our usage. It is said that a computer program has converged to a solution when it has found a solution with an error smaller than the tolerance.

For computing roots, we want an $x_r$ such that $f(x_r)$ is very close to 0. Therefore $|f(x)|$ is a possible choice for the measure of error since the smaller it is, the likelier we are to the root. If we assume that $x_i$ is the ith guess of an algorithm for finding a root, then $|x_{i-1} - x_i|$ is another possible choice for measuring error, since we expect the improvements between subsequent guesses to diminish as it approaches a solution.


Example:

Let the error be measured by e = $|f(x)|$ and tol be the acceptable level of error. The function $f(x) = x^2 + tol/2$ has no real roots. However, $|f(0)| = tol/2$ and is therefore acceptable as a solution for root finding program.

**19.3 - Bisection Method**

* The Intermediate Value Theorem (IVT) says that if f(x) is a continuous function between a and b, and the signs of f(a) and f(b) aren't the same, then there must be a c, such that a < c < b and f(c) = 0.

The bisection method uses the IVT iteratively to find roots.
* Let f(x) be a continuous function, and a and b be real scalar values such that a < b.
* Assume, without loss of generality, that f(a) > 0 and f(b) < 0.
* By the IVT, there must be a root on the open interval (a, b).
* Let m = (b + a) / 2, the midpoint between a and b.
* If f(m) = 0 or is close enough, then m is a root.
* If f(m) > 0, then m is an improvement on the left bound, a, and there is guaranteed to be a root on the open interval (m, b).
* If f(m) < 0, then m is an improvement on the right bound, b, and there is guaranteed to be a root on the open interval (a, m).
* The process can be repeated until the error is acceptably low

The square root of 2 can be computed as the root of the function $f(x) = x^2 - 2$. Starting at a = 0 and b = 2, we can use a function, my_bisection, to approximate the square root of 2 to a tolerance of $|f(x)| < 0.1$ and $|f(x) < 0.01$.

In [9]:
import numpy as np

def my_bisection(f, a, b, tol):
  # check to see if a and b bound a root
  if np.sign(f(a)) == np.sign(f(b)):
    raise Exception("The scalars a and b do not bound a root")

  # get a midpoint
  m = (a + b) / 2

  if np.abs(f(m)) < tol:
    # check to see if root
    return m
  elif np.sign(f(a)) == np.sign(f(m)):
    # case where m is an improvement on a
    # let a = m
    return my_bisection(f, m, b, tol)
  elif np.sign(f(b)) == np.sign(f(m)):
    # case where m is an improvement on b
    # let b = m
    return my_bisection(f, a, m, tol)

f = lambda x: x**2 - 2

r1 = my_bisection(f, 0, 2, 0.1)
print("r1 = ", r1)
r01 = my_bisection(f, 0, 2, 0.01)
print("r01 = ", r01)

print("f(r1) = ", f(r1))
print("f(r01) = ", f(r01))

r1 =  1.4375
r01 =  1.4140625
f(r1) =  0.06640625
f(r01) =  -0.00042724609375


As can be observed, a result is returned sooner with a larger tolerance. The result with the smaller tolerance, r01, is a much more accurate result when compared to the root.

**19.4 - Newton-Raphson Method**

Let f(x) be a smooth and continuous function and $x_r$ be an unknown root of f(x). Now assume that $x_0$ is a guess for $x_r$. Since $x_0$ is probably not the actual root, we are going to want to find an $x_1$ that is an improvement on $x_0$.

The linear approximation of f(x) around $x_0$ is $f(x)$ is approximately equal to $f(x_0) + f'(x_0)(x - x_0)$.
* Written generally, a Newton step computed an improved geuss, $x_i$, using a previous guess $x_{i-1}$, and is given by the following equation: $x_i = x_{i-1} - g(x_{i-1}) / g'(x_{i-1})$.
* The Newton-Raphson Method of finding roots iterates Newton steps from $x_0$ until the error is less than the tolerance.

Using the same example from the bisection method, we will compute the root using the Newton-Raphson Method.

In [11]:
import numpy as np

f = lambda x: x**2 - 2
f_prime = lambda x: 2*x
netwon_raphson = 1.4 - (f(1.4)) / (f_prime(1.4))

print("newton_raphson = ", netwon_raphson)
print("sqrt(2) = ", np.sqrt(2))

newton_raphson =  1.4142857142857144
sqrt(2) =  1.4142135623730951


The following is a function where the output is an estimation of the root f, f is a function object f(x), df is a function object to f'(x), x0 is an initial guess, and tol is the error tolerance. The error measurement should be |f(x)|.

In [12]:
def my_newton(f, df, x0, tol):
  # uses newon raphson method
  # recursive implementation
  if abs(f(x0)) < tol:
    return x0
  else:
    return my_newton(f, df, x0 - f(x0)/df(x0), tol)

We are now going to use it to compute square root of 2 to within tolerance 1e-6 starting at x0 = 1.5.

In [15]:
estimate = my_newton(f, f_prime, 1.5, 1e-6)
print("estimate = ", estimate)
print("sqrt(2) = ", np.sqrt(2))

estimate =  1.4142135623746899
sqrt(2) =  1.4142135623730951


If $x_0$ is close to $x_r$, then it can be proven that, in general, the Newton-Raphson method converges to $x_r$ much faster than the bisection method. Since $x_r$ is unknown, though, there is no way to know if the initial guess is close enough to the root to get this behavior unless some special information about the function is known ahead of time. Another limitation comes when the derivative at a guess is close to 0. Then the Newton step will be very large and probably lead away from the root. Depending on the behavior of the function derivate between $x_0$ and $x_r$, the Newton-Raphson method may converge to a different root than $x_r$ that may not be useful for our intention.